### PR TITLE
UNR-811 Spatial Switch Has Authority

### PR DESCRIPTION
**Pull request by Vix2021**
#### Description
Added a SpatialSwitchHasAuthority function that mimics the SpatialSitchHasAuthority Blueprint macro. 

[Top node is the new function when used on a Blueprint and bottom node is the existing Blueprint macro]

image

The difference in usability is that the existing node cannot be added to a Blueprint that does not derive from Actor whereas the new node can be added but gives an error if the Blueprint does not derive from Actor.

image

#### Release note
- Added `SpatialSwitchHasAuthority` function to differentiate authoritative server, non-authoritative server, and clients. This can be called in code or used in blueprints that derive from actor.

#### Tests
Added automated tests to SpatialAuthorityTest

How can this be verified by QA? Run  automated test: SpatialAuthorityTest

#### Documentation
Updated the GDK Multiserver document

### CODE REVIEW

**Comment by cmsmithio** [comment]
Thanks for putting this up.  A couple thoughts:

1) Is there a way to distinguish between this and the blueprint version of the node for teams that use it?  It would be nice if we could deprecate the Blueprint version and have that visualized in UnrealEd, but only if that functionality already exists.

2) I think we can drop the client auth output pin (I think this would only be relevant in non-Spatial paths, in which case this node shouldn't be used).  Michael thoughts?

**Comment by Vix2021** [comment]
> Thanks for putting this up. A couple thoughts:
> 
> 1. Is there a way to distinguish between this and the blueprint version of the node for teams that use it?  It would be nice if we could deprecate the Blueprint version and have that visualized in UnrealEd, but only if that functionality already exists.
> 2. I think we can drop the client auth output pin (I think this would only be relevant in non-Spatial paths, in which case this node shouldn't be used).  Michael thoughts?

Thanks for the feedback, in answer to your points:
1. It is easy to change the name of the existing Blueprint node to include DEPRECATED and also set the colour to red so that there is a clear difference between the nodes:
image

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include "SpatialAuthorityTestReplicatedActor.h"
 #include "SpatialFunctionalTestFlowController.h"
 #include "SpatialGDK/Public/EngineClasses/SpatialNetDriver.h"
+#include <SpatialGDK\Public\Utils\SpatialStatics.h>
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:13]: include formatting is inconsistent

**Review by m-samiec — APPROVED** [review]
Nice! Please add a changelog

**Comment by m-samiec** [comment]
> 2\. I think we can drop the client auth output pin (I think this would only be relevant in non-Spatial paths, in which case this node shouldn't be used).  Michael thoughts?

I think this is okay actually - if the client spawns something locally it would be client auth. Otherwise it'd be difficult to know what to return.

**Review by cmsmithio — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 1 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp]: Fixed the formatting.

**Comment by Vix2021** [comment]
> Nice! Please add a changelog

Added changelog

**Pull request merged**